### PR TITLE
Add option to enable opengl support in QCustomPlot

### PR DESCRIPTION
**Pull request by DonOregano**
Specify library name and version:  **qcustomplot/2.1.0**

This PR adds an option to enable opengl support in the QCustomPlot recipe. In the current recipe there is a TODO that this should be done, and this PR fixes that TODO. The default is to have opengl support disabled, so as to not change the behaviour for current users.
---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by DonOregano** [comment]
Something broke in the Visual Studio build. Will close and reopen this PR to trigger CI rebuild. (This is what I was told to do in another recipe I was working on).

**Comment by DonOregano** [comment]
@SpaceIm, looking at the history for qcustomplot conanfile it looks like you have been involved in the creation of this recipe. Can you help me understand what is going wrong here? There is something about a missing dependency in one of the visual studio builds.

**Comment by SpaceIm** [comment]
I guess it's not directly related to qcustomplot but to a new transitive dependency in qt dependency graph.

**Comment by SpaceIm** [comment]
/ping @jgsogo, qt5 recipe has been updated recently (https://github.com/conan-io/conan-center-index/pull/11531), and I don't see any other PR afterwards which could change its package ids, so maybe few binaries have not been uploaded to artifactory?

**Comment by DonOregano** [comment]
Thanks for looking into this! It did seem like something out of my control...

File: recipes/qcustomplot/all/conanfile.py [hunk 1]
```diff
@@ -76,7 +78,8 @@ def _patch_sources(self):
     def _configure_cmake(self):
         cmake = CMake(self)
         cmake.definitions["QT_VERSION"] = self.deps_cpp_info["qt"].version
-        # TODO: add an option to enable QCUSTOMPLOT_USE_OPENGL
+        if self.options.opengl:
+            cmake.definitions["QCUSTOMPLOT_USE_OPENGL"] = 1
```
Comment (SpaceIm) [inline recipes/qcustomplot/all/conanfile.py:82]: What do you expect from this CMake definition without a modification in the CMakeLists? Indeed, there is no QCUSTOMPLOT_USE_OPENGL option currently.

File: recipes/qcustomplot/all/conanfile.py [hunk 2]
```diff
@@ -18,10 +18,12 @@ class QcustomplotConan(ConanFile):
     options = {
         "shared": [True, False],
         "fPIC": [True, False],
+        "opengl": [True, False],
```
Comment (SpaceIm) [inline recipes/qcustomplot/all/conanfile.py:21]:
```suggestion
        "with_opengl": [True, False],
```

File: recipes/qcustomplot/all/conanfile.py [hunk 3]
```diff
@@ -18,10 +18,12 @@ class QcustomplotConan(ConanFile):
...
     }
     default_options = {
         "shared": False,
         "fPIC": True,
+        "opengl": False,
```
Comment (SpaceIm) [inline recipes/qcustomplot/all/conanfile.py:26]:
```suggestion
        "with_opengl": False,
```

File: recipes/qcustomplot/all/conanfile.py [hunk 4]
```diff
@@ -76,7 +78,8 @@ def _patch_sources(self):
     def _configure_cmake(self):
         cmake = CMake(self)
         cmake.definitions["QT_VERSION"] = self.deps_cpp_info["qt"].version
-        # TODO: add an option to enable QCUSTOMPLOT_USE_OPENGL
+        if self.options.opengl:
```
Comment (SpaceIm) [inline recipes/qcustomplot/all/conanfile.py:81]:
```suggestion
        if self.options.with_opengl:
```

File: recipes/qcustomplot/all/conanfile.py [hunk 5]
```diff
@@ -95,4 +98,6 @@ def package_info(self):
         self.cpp_info.libs = ["qcustomplot" + postfix]
         if self.options.shared:
             self.cpp_info.defines.append("QCUSTOMPLOT_USE_LIBRARY")
+        if self.options.opengl:
```
Comment (SpaceIm) [inline recipes/qcustomplot/all/conanfile.py:101]:
```suggestion
        if self.options.with_opengl:
```

File: recipes/qcustomplot/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [inline recipes/qcustomplot/all/conanfile.py:82]:
```suggestion
        cmake.definitions["QCUSTOMPLOT_USE_OPENGL"] = self.options.with_opengl
```

File: recipes/qcustomplot/all/CMakeLists.txt [hunk 6]
```diff
@@ -32,6 +32,11 @@ option(QCUSTOMPLOT_USE_OPENGL "Use OpenGL" OFF)
 
 if(QCUSTOMPLOT_USE_OPENGL)
     target_compile_definitions(${PROJECT_NAME} PUBLIC QCUSTOMPLOT_USE_OPENGL)
+	target_link_libraries(${PROJECT_NAME} PUBLIC Qt6::OpenGL)
+	if (MSVC)
+	    target_link_libraries(${PROJECT_NAME} PRIVATE OpenGL32)
+	endif()
+	#TODO: Other platforms may need other libraries linked in
```
Comment (SpaceIm) [inline recipes/qcustomplot/all/CMakeLists.txt:39]: Not like this, put this logic in `find_package()` blocks.

- And Qt5?
- OpenGL32 should come for free from `Qt6::OpenGL` target

Comment (DonOregano) [reply recipes/qcustomplot/all/CMakeLists.txt:39]: Oops... Will try to improve these....

**Comment by SpaceIm** [comment]
You should check in `validate()` that `opengl` option of qt recipe is not `no`.

**Comment by DonOregano** [comment]
> You should check in `validate()` that `opengl` option of qt recipe is not `no`.

Okay, I can do that. I do however have questions about how to handle opengl32 linking on windows specifically. I have been reading https://doc.qt.io/qt-6/windows-graphics.html, and from what I understand, unless the code uses QOpenGLFunctions (which QCustomPlot does not) I have to link to OpenGL32.lib on Windows to get things to work. Another alternative would be to require that `opengl` in Qt is `desktop`, which I believe would have the same result.

To summarize the Qt opengl thing I believe it works this way: If you specify "dynamic" you have to call opengl stuff through the QOpenGLFunctions class, and Qt will magically use the appropriate opengl library. If you specify "desktop" you will get linked to OpenGL32 (through QtX::OpenGL). If you specify "dynamic" you can still add linking to OpenGL32.lib explicitly, but I think this means exactly the same thing as using "desktop".

I would appreciate guidance on how to do this in the conan recipe, in a way that provides the fewest surprises to users and is maintainable...

Another solution would be to patch QCustomPlot according to this forum post https://www.qcustomplot.com/index.php/support/forum/1281, to make "dynamic" work, but that does seem a bit too invasive.

**Comment by DonOregano** [comment]
Since I didn't hear anything about my comment above I chose to just improve my initial implementation to get a reaction :-D

**Comment by jgsogo** [comment]
👍  It looks like that was the issue. Now build has passed. Time for reviews!

Thanks!

**Review by prince-chrismc — APPROVED** [review]

**Review by jgsogo — APPROVED** [review]

**Pull request merged**